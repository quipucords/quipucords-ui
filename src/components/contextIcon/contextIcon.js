import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '@patternfly/react-core';
import {
  CheckCircleIcon,
  ClipboardCheckIcon,
  CrosshairsIcon,
  DisconnectedIcon,
  DownloadIcon,
  ErrorCircleOIcon,
  ExclamationTriangleIcon,
  IdCardIcon,
  InfoCircleIcon,
  OpenshiftIcon,
  PencilAltIcon,
  PficonNetworkRangeIcon,
  PficonSatelliteIcon,
  PficonVcenterIcon,
  TrashIcon,
  UnknownIcon,
  UserIcon,
  IconSize
} from '@patternfly/react-icons';
import {
  global_Color_dark_100 as gray,
  global_success_color_100 as green,
  global_warning_color_100 as yellow,
  global_danger_color_100 as red
} from '@patternfly/react-tokens';

/**
 * Context icon colors, for consistency
 *
 * @type {{ red: object, gray: object, green: object, yellow: object }}
 */
const ContextIconColors = {
  gray,
  green,
  yellow,
  red
};

/**
 * Context icon variants
 *
 * @type {{paused: string, unreachable: string, sources: string, created: string, idCard: string, pending: string,
 *     completed: string, failed: string, error: string, pencil: string, network: string, trash: string, unknown: string,
 *     running: string, canceled: string, download: string, scans: string, success: string, cancelled: string,
 *     warning: string, vcenter: string, satellite: string, user: string, info: string}}
 */
const ContextIconVariant = {
  completed: 'success',
  success: 'success',
  error: 'failed',
  failed: 'failed',
  canceled: 'failed',
  cancelled: 'failed',
  created: 'pending',
  pending: 'pending',
  running: 'pending',
  paused: 'warning',
  warning: 'warning',
  download: 'download',
  idCard: 'idCard',
  info: 'info',
  network: 'network',
  openshift: 'openshift',
  pencil: 'pencil',
  satellite: 'satellite',
  scans: 'scans',
  sources: 'sources',
  trash: 'trash',
  unknown: 'unknown',
  unreachable: 'unreachable',
  user: 'user',
  vcenter: 'vcenter'
};

/**
 * Emulate pf icon sizing for custom SVGs
 *
 * @param {string} size
 * @returns {string} em measurement
 */
const svgSize = size => {
  if (!Number.isNaN(Number.parseFloat(size))) {
    return size;
  }

  switch (size) {
    case 'md':
      return '1.5em';
    case 'lg':
      return '2em';
    case 'xl':
      return '3em';
    case 'sm':
    default:
      return '1em';
  }
};

/**
 * Return an icon from context/symbol
 *
 * @param {object} props
 * @param {string} props.symbol
 * @param {object} props.props
 * @returns {React.ReactNode}
 */
const ContextIcon = ({ symbol, ...props }) => {
  switch (symbol) {
    case ContextIconVariant.download:
      return <DownloadIcon {...props} />;
    case ContextIconVariant.failed:
      return <ErrorCircleOIcon {...{ ...{ color: red.value }, ...props }} />;
    case ContextIconVariant.idCard:
      return <IdCardIcon {...{ ...{ color: gray.value }, ...props }} />;
    case ContextIconVariant.info:
      return <InfoCircleIcon {...props} />;
    case ContextIconVariant.network:
      return <PficonNetworkRangeIcon {...props} />;
    case ContextIconVariant.openshift:
      return <OpenshiftIcon {...props} />;
    case ContextIconVariant.warning:
      return <ExclamationTriangleIcon {...{ ...{ color: yellow.value }, ...props }} />;
    case ContextIconVariant.pencil:
      return <PencilAltIcon {...props} />;
    case ContextIconVariant.pending:
      const updatedSize = { style: { display: 'inline-block' } };

      if (props.size) {
        updatedSize.size = undefined;
        updatedSize.style.height = svgSize(props.size);
        updatedSize.style.width = svgSize(props.size);
      }

      return <Spinner isSVG {...{ ...props, ...updatedSize }} />;
    case ContextIconVariant.satellite:
      return <PficonSatelliteIcon {...props} />;
    case ContextIconVariant.scans:
      return <ClipboardCheckIcon {...props} />;
    case ContextIconVariant.sources:
      return <CrosshairsIcon {...props} />;
    case ContextIconVariant.success:
      return <CheckCircleIcon {...{ ...{ color: green.value }, ...props }} />;
    case ContextIconVariant.trash:
      return <TrashIcon {...props} />;
    case ContextIconVariant.unreachable:
      return <DisconnectedIcon {...{ ...{ color: red.value }, ...props }} />;
    case ContextIconVariant.user:
      return <UserIcon {...props} />;
    case ContextIconVariant.vcenter:
      return <PficonVcenterIcon {...props} />;
    case ContextIconVariant.unknown:
    default:
      return <UnknownIcon {...props} />;
  }
};

/**
 * Prop types
 *
 * @type {{symbol: string}}
 */
ContextIcon.propTypes = {
  symbol: PropTypes.oneOf([...Object.values(ContextIconVariant)]),
  size: PropTypes.oneOf([...Object.values(IconSize)])
};

/**
 * Default props
 *
 * @type {{symbol: null}}
 */
ContextIcon.defaultProps = {
  symbol: null,
  size: IconSize.sm
};

export { ContextIcon as default, ContextIcon, ContextIconColors, ContextIconVariant };
