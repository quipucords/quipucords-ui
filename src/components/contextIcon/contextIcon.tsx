/**
 * Provides a React component for rendering context-specific icons based on status or type, using PatternFly icons
 * and styling tokens for color consistency. Supports various icon sizes and custom props for additional configuration.
 * @module contextIcon
 */
import * as React from 'react';
import { Spinner } from '@patternfly/react-core';
import {
  AnsibleTowerIcon,
  CheckCircleIcon,
  ClipboardCheckIcon,
  CloudSecurityIcon,
  CrosshairsIcon,
  DisconnectedIcon,
  DownloadIcon,
  ExclamationCircleIcon,
  ExclamationTriangleIcon,
  IdCardIcon,
  InfoCircleIcon,
  OpenshiftIcon,
  PencilAltIcon,
  PficonNetworkRangeIcon,
  PficonSatelliteIcon,
  PficonVcenterIcon,
  TrashIcon,
  UnknownIcon,
  UserIcon
} from '@patternfly/react-icons';
import {
  global_Color_dark_100 as gray,
  global_success_color_100 as green,
  global_warning_color_100 as yellow,
  global_danger_color_100 as red
} from '@patternfly/react-tokens';

/**
 * Context icon colors, for consistency
 *
 * @type {{ red: object, gray: object, green: object, yellow: object }}
 */
type colorToken = {
  name: string;
  value: string;
  var: string;
};

const ContextIconColors: {
  gray: colorToken;
  green: colorToken;
  yellow: colorToken;
  red: colorToken;
} = {
  gray,
  green,
  yellow,
  red
};

const ContextIconVariant: { [key: string]: string } = {
  completed: 'success',
  success: 'success',
  error: 'failed',
  failed: 'failed',
  canceled: 'failed',
  cancelled: 'failed',
  created: 'pending',
  pending: 'pending',
  running: 'pending',
  paused: 'warning',
  warning: 'warning',
  download: 'download',
  idCard: 'idCard',
  info: 'info',
  network: 'network',
  openshift: 'openshift',
  pencil: 'pencil',
  satellite: 'satellite',
  scans: 'scans',
  sources: 'sources',
  trash: 'trash',
  unknown: 'unknown',
  unreachable: 'unreachable',
  user: 'user',
  vcenter: 'vcenter',
  ansible: 'ansible',
  acs: 'acs'
};

const ContextIcon: React.FC<{
  symbol: string;
  size?: 'sm' | 'md' | 'lg' | 'xl';
}> = ({ symbol, size, ...props }) => {
  switch (symbol) {
    case ContextIconVariant.ansible:
      return <AnsibleTowerIcon {...props} />;
    case ContextIconVariant.download:
      return <DownloadIcon {...props} />;
    case ContextIconVariant.failed:
      return <ExclamationCircleIcon {...{ ...{ color: red.value }, ...props }} />;
    case ContextIconVariant.idCard:
      return <IdCardIcon {...{ ...{ color: gray.value }, ...props }} />;
    case ContextIconVariant.info:
      return <InfoCircleIcon {...props} />;
    case ContextIconVariant.network:
      return <PficonNetworkRangeIcon {...props} />;
    case ContextIconVariant.openshift:
      return <OpenshiftIcon {...props} />;
    case ContextIconVariant.acs:
      return <CloudSecurityIcon {...props} />;
    case ContextIconVariant.warning:
      return <ExclamationTriangleIcon {...{ ...{ color: yellow.value }, ...props }} />;
    case ContextIconVariant.pencil:
      return <PencilAltIcon {...props} />;
    case ContextIconVariant.pending:
      return <Spinner size={size || 'sm'} {...props} />;
    case ContextIconVariant.satellite:
      return <PficonSatelliteIcon {...props} />;
    case ContextIconVariant.scans:
      return <ClipboardCheckIcon {...props} />;
    case ContextIconVariant.sources:
      return <CrosshairsIcon {...props} />;
    case ContextIconVariant.success:
      return <CheckCircleIcon {...{ ...{ color: green.value }, ...props }} />;
    case ContextIconVariant.trash:
      return <TrashIcon {...props} />;
    case ContextIconVariant.unreachable:
      return <DisconnectedIcon {...{ ...{ color: red.value }, ...props }} />;
    case ContextIconVariant.user:
      return <UserIcon {...props} />;
    case ContextIconVariant.vcenter:
      return <PficonVcenterIcon {...props} />;
    case ContextIconVariant.unknown:
    default:
      return <UnknownIcon {...props} />;
  }
};

export { ContextIcon as default, ContextIcon, ContextIconColors, ContextIconVariant };
