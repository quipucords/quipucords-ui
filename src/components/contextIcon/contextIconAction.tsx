/**
 * Renders action-specific icons (play, pause, stop, redo) based on the provided status symbol,
 * leveraging PatternFly icons. Designed for intuitive control and feedback in UI interactions.
 * @module contextIconAction
 */
import React from 'react';
import { PauseIcon, PlayIcon, RedoIcon, StopIcon } from '@patternfly/react-icons';

const ContextIconActionVariant = {
  completed: 'completed',
  failed: 'failed',
  canceled: 'failed',
  cancelled: 'failed',
  pending: 'pending',
  created: 'running',
  running: 'running',
  paused: 'paused',
  play: 'play'
};

const ContextIconAction: React.FC<{ symbol?: string }> = ({ symbol, ...props }) => {
  switch (symbol) {
    case ContextIconActionVariant.completed:
    case ContextIconActionVariant.failed:
      return <RedoIcon {...props} />;
    case ContextIconActionVariant.running:
      return <PauseIcon {...props} />;
    case ContextIconActionVariant.paused:
      return <PlayIcon {...props} />;
    case ContextIconActionVariant.pending:
      return <StopIcon {...props} />;
    default:
      return <PlayIcon {...props} />;
  }
};

export { ContextIconAction as default, ContextIconAction, ContextIconActionVariant };
