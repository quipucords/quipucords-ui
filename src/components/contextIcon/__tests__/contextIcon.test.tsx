import React from 'react';
import { shallowComponent } from '../../../../config/jest.setupTests';
import { ContextIcon, ContextIconVariant } from '../contextIcon';

describe('ContextIcon', () => {
  test.each(Object.keys(ContextIconVariant))('Should render icon %s', async iconName => {
    const props = { symbol: iconName };
    const component = await shallowComponent(<ContextIcon {...props} />);
    expect(component).toMatchSnapshot();
  });
});
