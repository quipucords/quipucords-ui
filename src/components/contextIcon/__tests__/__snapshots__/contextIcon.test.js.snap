// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextIcon Component should export, and handle icon variants: icon variant, canceled 1`] = `
<ErrorCircleOIcon
  color="#c9190b"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, cancelled 1`] = `
<ErrorCircleOIcon
  color="#c9190b"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, completed 1`] = `
<CheckCircleIcon
  color="#3e8635"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, created 1`] = `
<Spinner
  isSVG={true}
  style={
    Object {
      "display": "inline-block",
      "height": "1em",
      "width": "1em",
    }
  }
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, download 1`] = `
<DownloadIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, error 1`] = `
<ErrorCircleOIcon
  color="#c9190b"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, failed 1`] = `
<ErrorCircleOIcon
  color="#c9190b"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, idCard 1`] = `
<IdCardIcon
  color="#151515"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, info 1`] = `
<InfoCircleIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, network 1`] = `
<PficonNetworkRangeIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, openshift 1`] = `
<OpenshiftIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, paused 1`] = `
<ExclamationTriangleIcon
  color="#f0ab00"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, pencil 1`] = `
<PencilAltIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, pending 1`] = `
<Spinner
  isSVG={true}
  style={
    Object {
      "display": "inline-block",
      "height": "1em",
      "width": "1em",
    }
  }
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, running 1`] = `
<Spinner
  isSVG={true}
  style={
    Object {
      "display": "inline-block",
      "height": "1em",
      "width": "1em",
    }
  }
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, satellite 1`] = `
<PficonSatelliteIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, scans 1`] = `
<ClipboardCheckIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, sources 1`] = `
<CrosshairsIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, success 1`] = `
<CheckCircleIcon
  color="#3e8635"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, trash 1`] = `
<TrashIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, unknown 1`] = `
<UnknownIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, unreachable 1`] = `
<DisconnectedIcon
  color="#c9190b"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, user 1`] = `
<UserIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, vcenter 1`] = `
<PficonVcenterIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: icon variant, warning 1`] = `
<ExclamationTriangleIcon
  color="#f0ab00"
  noVerticalAlign={false}
  size="sm"
/>
`;

exports[`ContextIcon Component should export, and handle icon variants: variants 1`] = `
Object {
  "canceled": "failed",
  "cancelled": "failed",
  "completed": "success",
  "created": "pending",
  "download": "download",
  "error": "failed",
  "failed": "failed",
  "idCard": "idCard",
  "info": "info",
  "network": "network",
  "openshift": "openshift",
  "paused": "warning",
  "pencil": "pencil",
  "pending": "pending",
  "running": "pending",
  "satellite": "satellite",
  "scans": "scans",
  "sources": "sources",
  "success": "success",
  "trash": "trash",
  "unknown": "unknown",
  "unreachable": "unreachable",
  "user": "user",
  "vcenter": "vcenter",
  "warning": "warning",
}
`;

exports[`ContextIcon Component should render a basic component with default props: basic 1`] = `
<UnknownIcon
  color="currentColor"
  noVerticalAlign={false}
  size="sm"
/>
`;
